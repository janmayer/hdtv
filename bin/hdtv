#!/usr/bin/env python
# -*- coding: utf-8 -*-
#-------------------------------------------------------------------------
# This is the main HDTV application.
#-------------------------------------------------------------------------

from __future__ import print_function

import sys
import os
import glob

import argparse

# Reset command line arguments so that ROOT does not stumble about them
hdtv_args = sys.argv[1:]
sys.argv = [sys.argv[0]]

# Get config and data directory
legacypath = os.path.join(os.environ["HOME"], ".hdtv")
configpath = os.getenv("HDTV_USER_PATH",
    legacypath if os.path.isdir(legacypath) else
    os.path.join(os.getenv("XDG_CONFIG_HOME", 
        os.path.join(os.environ["HOME"], ".config")), "hdtv"))
datapath = os.getenv("HDTV_USER_PATH",
    legacypath if os.path.isdir(legacypath) else
    os.path.join(os.getenv("XDG_DATA_HOME", 
        os.path.join(os.environ["HOME"], ".local", "share")), "hdtv"))

for path in [datapath, configpath]:
    try:
        os.makedirs(path)
    except OSError:
        pass

if not os.access(datapath, os.W_OK) and os.access(configpath, os.W_OK):
    print("Could not access data path " + datapath +
          ", falling back to " + configpath, file=sys.stderr)
    datapath = configpath

if not os.access(configpath, os.R_OK):
    print("Could not access config path " + configpath, file=sys.stderr)

os.environ["HDTV_USER_PATH"] = configpath
sys.path.append(configpath)
sys.path.append(configpath + "/plugins")

# Check if we are inside the hdtv source directory and append
hdtvpath = os.sep.join(sys.path[0].split(os.sep)[:-1])
if os.path.exists(os.path.join(hdtvpath, 'hdtv')):
    sys.path.insert(1, hdtvpath)

# This needs to happen before "import ROOT" is called the first time â€“
# if not, there will be a header not found error
# for *some* ROOT versions and install methods
import hdtv.rootext
hdtv.rootext.UpdateRootIncludePath()

from hdtv._version import get_versions
__version__ = get_versions()['version']
del get_versions
# Command line arguments
parser = argparse.ArgumentParser()
parser.add_argument("-b", "--batch", dest="batchfile",
    help="Open and execute HDTV batchfile")
parser.add_argument("-e", "--execute", dest="commands",
    help="Execute HDTV command(s)")
parser.add_argument("-v", "--version", action="version",
    help="Show HDTV Version",
    version="HDTV {}".format(__version__))
parser.add_argument("--rebuild-usr", action='store_true', dest='rebuildusr',
    help='Rebuild ROOT-loadable libraries for the current user')
parser.add_argument("--rebuild-sys", action='store_true', dest='rebuildsys',
    help='Rebuild ROOT-loadable libraries for all users')
args = parser.parse_args(hdtv_args)

if args.rebuildusr:
    import hdtv.rootext.dlmgr
    hdtv.rootext.dlmgr.RebuildLibraries(hdtv.rootext.dlmgr.usrdir)
if args.rebuildsys:
    import hdtv.rootext.dlmgr
    hdtv.rootext.dlmgr.RebuildLibraries(hdtv.rootext.dlmgr.sysdir)

if args.rebuildusr or args.rebuildsys:
    sys.exit(0)

# Import core modules
import hdtv.cmdline
import hdtv.session
import hdtv.ui


hdtv.cmdline.ReadReadlineInit(configpath + "/inputrc")
hdtv.cmdline.SetReadlineHistory(datapath + "/hdtv_history")
hdtv.cmdline.SetInteractiveDict(locals())
spectra = hdtv.session.Session()

# Import core plugins
import hdtv.plugins.textInterface
import hdtv.plugins.ls
import hdtv.plugins.run
import hdtv.plugins.specInterface
import hdtv.plugins.fitInterface
import hdtv.plugins.calInterface
import hdtv.plugins.matInterface
import hdtv.plugins.rootInterface
import hdtv.plugins.config
import hdtv.plugins.fitlist
import hdtv.plugins.fittex
import hdtv.plugins.fitmap
import hdtv.plugins.dblookup
import hdtv.plugins.peakfinder
import hdtv.plugins.printing


hdtv.ui.msg("HDTV - Nuclear Spectrum Analysis Tool")


# Execute startup.py for user configuration in python
try:
    import startup
except ImportError:
    hdtv.ui.debug("No startup.py file")

# Execute startup.hdtv and startup.hdtv.d/*.hdtv
# for user configuration in "hdtv" language
startup_d_hdtv = [configpath + os.sep + "startup.hdtv"] + glob.glob(
    configpath + os.sep + "startup.hdtv.d" + os.sep + "*.hdtv")

for startup_hdtv in startup_d_hdtv:
    try:
        if os.path.exists(startup_hdtv):
            hdtv.cmdline.command_line.ExecCmdfile(startup_hdtv)
    except IOError as msg:
        hdtv.ui.error("Error reading %s: %s" % (startup_hdtv, msg))

# Execute batchfile given on command line
try:
    if args.batchfile is not None:
        hdtv.cmdline.command_line.ExecCmdfile(args.batchfile)
except IOError as msg:
    hdtv.ui.msg("Error reading %s: %s" % (args.batchfile, msg))

if args.commands is not None:
    hdtv.cmdline.command_line.DoLine(args.commands)


# Go
hdtv.cmdline.MainLoop()
hdtv.plugins.rootInterface.r.rootfile = None
hdtv.cmdline.command_tree.SetDefaultLevel(1)
