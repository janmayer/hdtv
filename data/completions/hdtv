_hdtv() 
{
    local cur prev opts
    COMPREPLY=()
    cur="${COMP_WORDS[COMP_CWORD]}"
    prev="${COMP_WORDS[COMP_CWORD-1]}"
    opts="--batch --execute --help --version --rebuild-usr --rebuild-sys"

    if [[ ${prev} == --batch ]] || [[ ${prev} == -b ]]; then
        _filedir
    elif [[ ${cur} == -* ]] ; then
        COMPREPLY=( $(compgen -W "${opts}" -- ${cur}) )
        return 0
    fi
}
complete -F _hdtv hdtv
