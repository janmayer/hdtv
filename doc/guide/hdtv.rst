======
 HDTV
======

------------------------------
Nuclear Spectrum Analysis Tool
------------------------------

:Manual section: 1

SYNOPSIS
========

| **hdtv** [ **-h** ] [ **-b** *batchfile* ] [ **-e** *commands* ] [ **-v** ]

DESCRIPTION
===========

HDTV tries to provide functionality similar to the old TV program on top
of the ``ROOT``\(1) data analysis toolkit developed at CERN. The use of Python
gives HDTV much better scripting capabilities than TV. Also, since HDTV
consists of a number of modules that can in principle be used independently
of each other, HDTV is much easier to extend and customize. HDTV is written
in a mixture of C++ and Python, glued together using PyROOT.

OPTIONS
=======

--help, -h
    Show help message and exit

--batch *batchfile*, -b *batchfile*
    Open and execute hdtv batchfile

--execute *commands*, -e *commands*
    Execute hdtv command(s)

--version, -v
    Show HDTV Version

HDTV COMMAND LINE
=================

HDTV features an extensive command line interface. Besides a wide set of
HDTV commands, it is possible to use special prefixes:

:
    Execute a single Python command.

@
    Execute an HDTV batch file (or use the ``exec`` command).

!
    Execute a single shell command.

The command line is implemented using ``prompt_toolkit``. Common features,
such as tab-completion, history and history-search (using ``Ctrl-R``) are
available.

KEY BINDINGS
============

Note: Key bindings are case sensitive.

Window
------

u
    Update Viewport

l
    Toggle Log Scale

A
    Toggle Y Auto Scale

!
    Toggle Use Norm


x directions
^^^^^^^^^^^^

Space
    Set X Zoom Marker

x
    Expand X

Right, >
    Shift X Offset Right

Left, <
    Shift X Offset Left

Return, Enter
    Update

\|
    Set X Center at Cursor

1
    X Zoom In Around Cursor

0
    X Zoom Out Around Cursor

y direction
^^^^^^^^^^^

h
    Set Y Zoom Marker

y
    Expand Y

Up
    Shift Y Offset Up

Down
    Shift Y Offset Down

Z
    Y Zoom In Around Cursor

X
    Y Zoom Out Around Cursor


all directions
^^^^^^^^^^^^^^

e
    Expand

i
    Enter Edit Mode: Go To Position


Spec Interface
--------------

PageUp, N + p
    Show Prev

PageDown, N + n
    Show Next

=
    Refresh All

t
    Refresh Visible

n
    Enter Edit Mode: Show spectrum

a
    Enter Edit Mode: Activate spectrum


Fit Interface
-------------

b
    Set Marker bg

Minus + b
    Remove Marker bg

r
    Set Marker region

Minus + r
    Remove Marker region

p
    Set Marker peak

Minus + p
    Remove Marker peak

B
    Execute Fit (Background only)

F
    Execute Fit

Q
    QuickFit

Minus + B
    Clear Fit (Background only)

Minus + F
    Clear Fit

Plus + F
    Store Fit

D
    Show Decomposition

Minus + D
    Hide Decomposition

f + s
    Enter Edit Mode: Show Fit

f + a
    Enter Edit Mode: Activate Fit

f + p
    Show Prev

f + n
    Show Next

I
    Execute Integral


Matrix Interface
----------------

g
    Set Marker cut

c + g
    Set Marker cutregion

c + b
    Set Marker cutbg

Minus + c + g
    Remove Marker cutregion

Minus + c + b
    Remove Marker cutbg

Minus + C
    Clear Cut

Plus + C
    Store Cut

C
    Execute Cut

c + s
    Enter Edit Mode: Show Cut

c + a
    Enter Edit Mode: Activate Cut

c + p
    Show Prev

c + n
    Show Next

Tab
    Switch

FILES
=====

$HOME/.config/hdtv/startup.py
    Python script that gets executed during startup.

$HOME/.config/hdtv/startup.hdtv, $HOME/.config/hdtv/startup.hdtv.d/\*.hdtv
    Files containing HDTV commands that get executed during startup.

$HOME/.config/hdtv/plugins/
   Put plugins here.

$HOME/.local/share/hdtv/hdtv_history
    History of commands executed in HDTV.

HDTV supports the XDG Base Directory Specification, with the default paths
listed above. If the legacy directory ``$HOME/.hdtv`` exists, it is used
instead. It is also possible to manually set the directory using the
``$HDTV_USER_PATH`` environment variable.

BUGS
====

See https://gitlab.ikp.uni-koeln.de/staging/hdtv/issues/ for a list of all
currently open bugs and feature requests.

AUTHORS
=======

Jan Mayer (jan.mayer@ikp.uni-koeln.de),
Elena Hoemann (ehoemann@ikp.uni-koeln.de),
Oliver Papst (opapst@ikp.tu-darmstadt.de),
Nigel Warr (warr@ikp.uni-koeln.de),
Norbert Braun (n.braun@ikp.uni-koeln.de),
Tanja Kotthaus (t.kotthaus@ikp.uni-koeln.de),
Ralf Schulze (r.schulze@ikp.uni-koeln.de)

SEE ALSO
========

| ``hdtv-tutorial``\(1), ``root``\(1), ``python``\(1)

