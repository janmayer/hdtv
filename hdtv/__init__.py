import os

installdir = os.path.dirname(__file__)
datadir = os.path.join(installdir, "share")
