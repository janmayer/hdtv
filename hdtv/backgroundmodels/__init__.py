from .exponential import BackgroundModelExponential
from .polynomial import BackgroundModelPolynomial
from .interpolation import BackgroundModelInterpolation

# dictionary of available background models
BackgroundModels = dict()
BackgroundModels["exponential"] = BackgroundModelExponential
BackgroundModels["polynomial"] = BackgroundModelPolynomial
BackgroundModels["interpolation"] = BackgroundModelInterpolation
