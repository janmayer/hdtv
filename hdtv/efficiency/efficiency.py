# -*- coding: utf-8 -*-

# HDTV - A ROOT-based spectrum analysis software
#  Copyright (C) 2006-2009  The HDTV development team (see file AUTHORS)
#
# This file is part of HDTV.
#
# HDTV is free software; you can redistribute it and/or modify it
# under the terms of the GNU General Public License as published by the
# Free Software Foundation; either version 2 of the License, or (at your
# option) any later version.
#
# HDTV is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
# FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License
# for more details.
#
# You should have received a copy of the GNU General Public License
# along with HDTV; if not, write to the Free Software Foundation,
# Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA

import array
import string
from math import sqrt
from uncertainties import ufloat

from hdtv.util import TxtFile, Pairs
import hdtv.ui
from ROOT import TF1, TF2, TGraphErrors, TVirtualFitter


class _Efficiency(object):
    def __init__(self, num_pars=0, pars=None, norm=True):
        pars = pars or []

        self._numPars = num_pars
        self.parameter = pars
        self.fCov = [
            [None for j in range(self._numPars)] for i in range(self._numPars)
        ]  # Simple matrix replacement
        self._dEff_dP = list()
        self.TGraph = TGraphErrors()

        # Normalization factors
        self._doNorm = norm
        self.norm = 1.0
        self.TF1.FixParameter(0, self.norm)  # Normalization
        self.TF1.SetRange(0, 10000)  # Default range for efficiency function

        self._fitInput = Pairs(lambda x: ufloat(x, 0))

        #         if self.parameter: # Parameters were given
        #             map(lambda i: self.TF1.SetParameter(i + 1, self.parameter[i]), range(1, len(pars))) # Set initial parameters
        #         else:
        #             self.parameter = [None for i in range(1, self._numPars + 1)]
        #
        self.TF1.SetParName(0, "N")  # Normalization

        for i in range(0, num_pars):
            self._dEff_dP.append(None)
            if num_pars <= len(string.ascii_lowercase):
                self.TF1.SetParName(i + 1, string.ascii_lowercase[i])

    def _getParameter(self):
        """
        Get parameter of efficiency function
        """
        pars = list()
        for i in range(self._numPars):
            pars.append(self.TF1.GetParameter(i))

        return pars

    def _setParameter(self, pars):
        """
        Set parameter for efficiency function
        """
        for i in range(self._numPars):
            try:
                self.TF1.SetParameter(i, pars[i])
            except IndexError:
                self.TF1.SetParameter(i, 0)

    parameter = property(_getParameter, _setParameter)

    def __call__(self, E):
        value = self.value(E)
        try:
            error = self.error(E)
        except TypeError:
            error = None
        return ufloat(value, error)

    def _set_fitInput(self, fitPairs):

        self._fitInput = fitPairs

        for i in range(len(self._fitInput)):
            p = self._fitInput[i]
            try:
                e_nominal_value = p[0].nominal_value
                e_std_dev = p[0].std_dev
            except AttributeError:
                e_nominal_value = float(p[0])
                e_std_dev = 0.0

            try:
                eff_nominal_value = p[1].nominal_value
                eff_std_dev = p[1].std_dev
            except AttributeError:
                eff_nominal_value = float(p[1])
                eff_std_dev = 0.0

            self.TGraph.SetPoint(i, e_nominal_value, eff_nominal_value)
            self.TGraph.SetPointError(i, e_std_dev, eff_std_dev)

    def _get_fitInput(self):
        return self._fitInput

    fitInput = property(_get_fitInput, _set_fitInput)

    def fit(self, fitPairs=None, quiet=True):
        """
        Fit efficiency curve to values given by 'fitPairs' which should be a list
        of energy<->efficiency pairs. (See hdtv.util.Pairs())

        'energies' and 'efficiencies' may be a list of ufloats
        """
        # TODO: Unify this with the energy calibration fitter
        if fitPairs is not None:
            # self.fitInput = fitPairs
            self._fitInput = fitPairs

        E = array.array("d")
        delta_E = array.array("d")
        eff = array.array("d")
        delta_eff = array.array("d")
        EN = array.array("d")
        effN = array.array("d")

        #        map(energies.append(self.fitInput[0]), self.fitInput)
        #        map(efficiencies.append(self.fitInput[1]), self.fitInput)
        hasXerrors = False
        # Convert energies to array needed by ROOT
        try:
            list(map(lambda x: E.append(x[0].nominal_value), self._fitInput))
            list(map(lambda x: delta_E.append(x[0].std_dev), self._fitInput))
            list(map(lambda x: EN.append(0.0), self._fitInput))
            hasXerrors = True
        except AttributeError:  # energies does not seem to be ufloat list
            list(map(lambda x: E.append(x[0]), self._fitInput))
            list(map(lambda x: delta_E.append(0.0), self._fitInput))

        # Convert efficiencies to array needed by ROOT
        try:
            list(map(lambda x: eff.append(x[1].nominal_value), self._fitInput))
            list(map(lambda x: delta_eff.append(x[1].std_dev), self._fitInput))
            list(map(lambda x: effN.append(0.0), self._fitInput))
        except AttributeError:  # energies does not seem to be ufloat list
            list(map(lambda x: eff.append(x[1]), self._fitInput))
            list(map(lambda x: delta_eff.append(0.0), self._fitInput))

        # if fit has errors we first fit without errors to get good initial values
        # if hasXerrors == True:
        # print "Fit parameter without errors included:"
        # self.TGraphWithoutErrors = TGraphErrors(len(E), E, eff, EN, effN)
        # fitWithoutErrors = self.TGraphWithoutErrors.Fit(self.id, "SF")

        hdtv.ui.msg("Fit parameter with errors included:")

        # Preliminary normalization
        #        if self._doNorm:
        #            self.norm = 1 / max(efficiencies)
        #            for i in range(len(eff)):
        #                eff[i] *= self.norm
        #                delta_eff[i] *= self.norm

        # self.TF1.SetRange(0, max(E) * 1.1)
        # self.TF1.SetParameter(0, 1) # Unset normalization for fitting

        self.TGraph = TGraphErrors(len(E), E, eff, delta_E, delta_eff)

        # Do the fit
        fitopts = "0"  # Do not plot

        if hasXerrors:
            # We must use the iterative fitter (minuit) to take x errors
            # into account.
            fitopts += "F"
            hdtv.ui.info(
                "switching to non-linear fitter (minuit) for x error weighting"
            )
        if quiet:
            fitopts += "Q"

        fitopts += (
            "S"  # Additional fitinfo returned needed for ROOT5.26 workaround below
        )
        fitreturn = self.TGraph.Fit(self.id, fitopts)

        try:
            # Workaround for checking the fitstatus in ROOT 5.26 (TFitResultPtr
            # does not cast properly to int)
            fitstatus = fitreturn.Get().Status()
        except AttributeError:  # This is for ROOT <= 5.24, where fit returns an int
            fitstatus = int(fitreturn)

        if fitstatus != 0:
            # raise RuntimeError, "Fit failed"
            hdtv.ui.msg("Fit failed")

        #         # Final normalization
        #         if self._doNorm:
        #             self.normalize()

        # Get parameter
        for i in range(self._numPars):
            self.parameter[i] = self.TF1.GetParameter(i)

        # Get covariance matrix
        tvf = TVirtualFitter.GetFitter()
        ##        cov = tvf.GetCovarianceMatrix()
        for i in range(0, self._numPars):
            for j in range(0, self._numPars):
                self.fCov[i][j] = tvf.GetCovarianceMatrixElement(i, j)
        ##                 self.fCov[i][j] = cov[i * self._numPars + j]

        return self.parameter

    def normalize(self):
        # Normalize the efficiency funtion
        try:
            self.norm = 1.0 / self.TF1.GetMaximum(0.0, 0.0)
        except ZeroDivisionError:
            self.norm = 1.0

        self.TF1.SetParameter(0, self.norm)
        normfunc = TF2("norm_" + hex(id(self)), "[0]*y")
        normfunc.SetParameter(0, self.norm)
        self.TGraph.Apply(normfunc)

    def value(self, E):
        try:
            value = E.nominal_value
        except AttributeError:
            value = E

        return self.TF1.Eval(value)

    def error(self, E):
        """
        Calculate error using the covariance matrix via:

          delta_Eff = sqrt((dEff_dP[0], dEff_dP[1], ... dEff_dP[num_pars]) x cov x (dEff_dP[0], dEff_dP[1], ... dEff_dP[num_pars]))

        """
        try:
            value = E.nominal_value
        except AttributeError:
            value = E

        if not self.fCov or (len(self.fCov) != self._numPars):
            raise ValueError("Incorrect size of covariance matrix")

        res = 0.0

        # Do matrix multiplication
        for i in range(0, self._numPars):
            tmp = 0.0
            for j in range(0, self._numPars):
                tmp += self._dEff_dP[j](value, self.parameter) * self.fCov[i][j]

            res += self._dEff_dP[i](value, self.parameter) * tmp

        return sqrt(res)

    def loadPar(self, parfile):
        """
        Read parameter from file
        """
        vals = []

        file = TxtFile(parfile)
        file.read()

        for line in file.lines:
            vals.append(float(line))

        if len(vals) != self._numPars:
            raise RuntimeError("Incorrect number of parameters found in file")

        self.parameter = vals
        if self._doNorm:
            self.normalize()

    def loadCov(self, covfile):
        """
        Load covariance matrix from file
        """

        vals = []

        file = TxtFile(covfile)
        file.read()

        for line in file.lines:
            val_row = [float(s) for s in line.split()]
            if len(val_row) != self._numPars:
                raise RuntimeError("Incorrect format of parameter error file")
            vals.append(val_row)

        if len(vals) != self._numPars:
            raise RuntimeError("Incorrect format of parameter error file")

        self.fCov = vals

    def load(self, parfile, covfile=None):
        """
        Read parameter and covariance matrix from file
        """
        self.loadPar(parfile)

        if covfile:
            self.loadCov(covfile)

    def savePar(self, parfile):
        """
        Save parameter to file
        """
        file = TxtFile(parfile, "w")

        for p in self.parameter:
            file.lines.append(str(p))

        file.write()

    def saveCov(self, covfile):
        """
        Save covariance matrix to file
        """
        file = TxtFile(covfile, "w")

        for i in range(0, self._numPars):
            line = ""
            for j in range(0, self._numPars):
                line += str(self.fCov[i][j]) + " "
            file.lines.append(line)

        file.write()

    def save(self, parfile, covfile=None):
        """
        Save parameter and covariance matrix to files
        """
        # Write paramter
        self.savePar(parfile)

        # Write covariance matrix
        if covfile is not None:
            self.saveCov(covfile)
