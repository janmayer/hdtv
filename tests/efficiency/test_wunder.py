#!/usr/bin/env python

from __future__ import print_function

import hdtv.efficiency as eff

e = eff.WunderEff()

e.load("eff_wunder.par", "eff_wunder.cov")

for i in range(1, 10000, 100):
    print("Wunder-Efficiency @", i, " keV = ", e(i))
