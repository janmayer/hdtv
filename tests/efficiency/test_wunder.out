Wunder-Efficiency @ 1  keV =  5.742381(0)e-284
Wunder-Efficiency @ 101  keV =  3.45(12)e-2
Wunder-Efficiency @ 201  keV =  0.4328(58)
Wunder-Efficiency @ 301  keV =  0.8325(65)
Wunder-Efficiency @ 401  keV =  1.0465(72)
Wunder-Efficiency @ 501  keV =  1.1295(84)
Wunder-Efficiency @ 601  keV =  1.1401(93)
Wunder-Efficiency @ 701  keV =  1.1135(97)
Wunder-Efficiency @ 801  keV =  1.0693(98)
Wunder-Efficiency @ 901  keV =  1.0178(96)
Wunder-Efficiency @ 1001  keV =  0.9644(95)
Wunder-Efficiency @ 1101  keV =  0.9121(95)
Wunder-Efficiency @ 1201  keV =  0.8623(97)
Wunder-Efficiency @ 1301  keV =  0.816(10)
Wunder-Efficiency @ 1401  keV =  0.772(11)
Wunder-Efficiency @ 1501  keV =  0.732(12)
Wunder-Efficiency @ 1601  keV =  0.695(13)
Wunder-Efficiency @ 1701  keV =  0.661(14)
Wunder-Efficiency @ 1801  keV =  0.630(15)
Wunder-Efficiency @ 1901  keV =  0.601(16)
Wunder-Efficiency @ 2001  keV =  0.574(17)
Wunder-Efficiency @ 2101  keV =  0.550(18)
Wunder-Efficiency @ 2201  keV =  0.527(19)
Wunder-Efficiency @ 2301  keV =  0.505(19)
Wunder-Efficiency @ 2401  keV =  0.486(20)
Wunder-Efficiency @ 2501  keV =  0.467(20)
Wunder-Efficiency @ 2601  keV =  0.450(21)
Wunder-Efficiency @ 2701  keV =  0.434(21)
Wunder-Efficiency @ 2801  keV =  0.418(21)
Wunder-Efficiency @ 2901  keV =  0.404(22)
Wunder-Efficiency @ 3001  keV =  0.390(22)
Wunder-Efficiency @ 3101  keV =  0.377(22)
Wunder-Efficiency @ 3201  keV =  0.365(22)
Wunder-Efficiency @ 3301  keV =  0.354(21)
Wunder-Efficiency @ 3401  keV =  0.343(21)
Wunder-Efficiency @ 3501  keV =  0.332(21)
Wunder-Efficiency @ 3601  keV =  0.322(21)
Wunder-Efficiency @ 3701  keV =  0.313(20)
Wunder-Efficiency @ 3801  keV =  0.304(20)
Wunder-Efficiency @ 3901  keV =  0.295(19)
Wunder-Efficiency @ 4001  keV =  0.286(19)
Wunder-Efficiency @ 4101  keV =  0.278(19)
Wunder-Efficiency @ 4201  keV =  0.270(18)
Wunder-Efficiency @ 4301  keV =  0.263(17)
Wunder-Efficiency @ 4401  keV =  0.255(17)
Wunder-Efficiency @ 4501  keV =  0.248(16)
Wunder-Efficiency @ 4601  keV =  0.242(16)
Wunder-Efficiency @ 4701  keV =  0.235(15)
Wunder-Efficiency @ 4801  keV =  0.229(15)
Wunder-Efficiency @ 4901  keV =  0.223(14)
Wunder-Efficiency @ 5001  keV =  0.217(13)
Wunder-Efficiency @ 5101  keV =  0.211(13)
Wunder-Efficiency @ 5201  keV =  0.205(12)
Wunder-Efficiency @ 5301  keV =  0.200(12)
Wunder-Efficiency @ 5401  keV =  0.194(11)
Wunder-Efficiency @ 5501  keV =  0.189(10)
Wunder-Efficiency @ 5601  keV =  0.1842(98)
Wunder-Efficiency @ 5701  keV =  0.1793(92)
Wunder-Efficiency @ 5801  keV =  0.1745(87)
Wunder-Efficiency @ 5901  keV =  0.1699(81)
Wunder-Efficiency @ 6001  keV =  0.1654(76)
Wunder-Efficiency @ 6101  keV =  0.1610(71)
Wunder-Efficiency @ 6201  keV =  0.1568(65)
Wunder-Efficiency @ 6301  keV =  0.1526(60)
Wunder-Efficiency @ 6401  keV =  0.1486(55)
Wunder-Efficiency @ 6501  keV =  0.1446(51)
Wunder-Efficiency @ 6601  keV =  0.1408(46)
Wunder-Efficiency @ 6701  keV =  0.1370(42)
Wunder-Efficiency @ 6801  keV =  0.1334(38)
Wunder-Efficiency @ 6901  keV =  0.1298(34)
Wunder-Efficiency @ 7001  keV =  0.1263(30)
Wunder-Efficiency @ 7101  keV =  0.1230(27)
Wunder-Efficiency @ 7201  keV =  0.1197(23)
Wunder-Efficiency @ 7301  keV =  0.1164(21)
Wunder-Efficiency @ 7401  keV =  0.1133(18)
Wunder-Efficiency @ 7501  keV =  0.1103(17)
Wunder-Efficiency @ 7601  keV =  0.1073(16)
Wunder-Efficiency @ 7701  keV =  0.1044(16)
Wunder-Efficiency @ 7801  keV =  0.1016(16)
Wunder-Efficiency @ 7901  keV =  9.88(17)e-2
Wunder-Efficiency @ 8001  keV =  9.61(18)e-2
Wunder-Efficiency @ 8101  keV =  9.35(20)e-2
Wunder-Efficiency @ 8201  keV =  9.09(22)e-2
Wunder-Efficiency @ 8301  keV =  8.84(23)e-2
Wunder-Efficiency @ 8401  keV =  8.60(25)e-2
Wunder-Efficiency @ 8501  keV =  8.37(27)e-2
Wunder-Efficiency @ 8601  keV =  8.13(29)e-2
Wunder-Efficiency @ 8701  keV =  7.91(30)e-2
Wunder-Efficiency @ 8801  keV =  7.69(32)e-2
Wunder-Efficiency @ 8901  keV =  7.48(34)e-2
Wunder-Efficiency @ 9001  keV =  7.27(35)e-2
Wunder-Efficiency @ 9101  keV =  7.07(37)e-2
Wunder-Efficiency @ 9201  keV =  6.87(38)e-2
Wunder-Efficiency @ 9301  keV =  6.68(39)e-2
Wunder-Efficiency @ 9401  keV =  6.49(41)e-2
Wunder-Efficiency @ 9501  keV =  6.31(42)e-2
Wunder-Efficiency @ 9601  keV =  6.13(43)e-2
Wunder-Efficiency @ 9701  keV =  5.96(44)e-2
Wunder-Efficiency @ 9801  keV =  5.79(45)e-2
Wunder-Efficiency @ 9901  keV =  5.63(45)e-2
